src/init_parsing.jl

```diff
@@ -3,147 +3,185 @@ struct HeaderParsingError <: Exception
-                ncols = length(parsing_ctx.header)
+                ncols_actual = length(parsing_ctx.header)
                 while !Parsers.eof(code)
-                    res = Parsers.xparse(String, v, pos, length(v), options, Parsers.PosLen31)
+                    res = Parsers.xparse(String, row_bytes, pos, len, options, Parsers.PosLen31)
                     (tlen, code) = res.tlen, res.code
                     pos += tlen
-                    ncols += 1
+                    ncols_actual += 1
                 end
-                throw(HeaderParsingError("Error parsing header, there are more columns ($ncols) than provided types in schema ($(length(schema_input))) at $(lines_skipped_total+1):$(pos) (row:pos)."))
+                throw(HeaderParsingError("Error parsing header, there are more columns ($ncols_actual) than provided types in schema ($ncols) at $(lines_skipped_total+1):$(pos) (row:pos)."))
             end
         end
     elseif !should_parse_header
+        # The user didn't provide a schema nor the header and the file *doesn't* have a header,
+        # so we need to infer the number of columns from the first row and fill the schema with String.
         input_is_empty && return nothing
         # infer the number of columns from the first data row
-        s = chunking_ctx.newline_positions[1]
-        e = chunking_ctx.newline_positions[2]
-        v = @view chunking_ctx.bytes[s+1:e-1]
+        header_row_start = newlines[1]
+        header_row_end   = newlines[2]
+        row_bytes = @view chunking_ctx.bytes[header_row_start+1:header_row_end-1]
+
+        len = length(row_bytes)
         pos = 1
-        ignorerepeated && (pos = Parsers.checkdelim!(v, pos, length(v), options))
+        # `ignorerepeated` is used to implement fixed width column parsing. We need to skip over the initial
+        # delimiters to avoid getting an empty first value.
+        ignorerepeated && (pos = Parsers.checkdelim!(row_bytes, pos, len, options))
         code = Parsers.OK
         i = 1
         while !(Parsers.eof(code) || Parsers.newline(code))
-            res = Parsers.xparse(String, v, pos, length(v), options, Parsers.PosLen31)
+            res = Parsers.xparse(String, row_bytes, pos, len, options, Parsers.PosLen31)
             (tlen, code) = res.tlen, res.code
+
             !(Parsers.ok(code) || Parsers.sentinel(code)) && (throw(HeaderParsingError("Error parsing header for column $i at $(lines_skipped_total+1):$(pos) (row:pos).")))
+
+            push!(parsing_ctx.header, Symbol(string(input_args.default_colname_prefix, i)))
             pos += tlen
-            push!(parsing_ctx.header, Symbol(string(settings.default_colname_prefix, i)))
             i += 1
         end
-        _fill_schema!(schema, parsing_ctx.header, schema_input, settings)
+        _fill_schema!(schema, parsing_ctx.header, schema_input, input_args)
     else
+        # The user didn't provide a schema nor the header and the file *does* have a header,
```

Comment:
Maybe for a separate PR (as this change might make this one too hard to review)... 

I wonder if everyone of these `if/elseif/else` blocks should be a function? 

(which we force inline if that's a performance consideration, but i doubt it, given this is called once, right?)

e.g. this function would look like
```julia
    ...
    if schema_provided & header_provided
        schema_and_header!(args...)
    elseif !schema_provided & header_provided
        create_schema!(args...)
    elseif schema_provided & !header_provided
        create_header!(args...)
    elseif !should_parse_header
        create_schema_no_header!(args...)
    else
        create_schema_and_header!(args...)
    end
    !schema_provided && validate_schema(schema)
    skip_to_data!(args...)
    create_enum_schema!(parsing_ctx)
    return nothing
end
```

Reply:
Yeah I struggle to refactor these and this seems like a nice approach 👍